File: ULTRAIA/integracionesImplementacion/media_score.py
"""Puntuacion de calidad de media generada (loop de verificacion).

Uso:
  python media_score.py <resultado.json> [--verbose]

<resultado.json> es la salida de una generacion (p.ej. el manifest del pipeline,
  o la respuesta del Gen-Engine). Devuelve 0-100 y un veredicto PASS/FAIL.

Criterios (cada uno aporta puntos):
  - imagen: URL https de dominio conocido, con modelo explicito, sin logo.
  - audio:  formato mp3/wav, duracion > 0, loudnorm en rango.
  - video:  frames 1-8 o url de video, motion valida.
  - tts:    idioma soportado, voz edge-tts conocida.
  - director: plan con script, images >= 1, language en los soportados.
"""
from __future__ import annotations

import json
import sys
from pathlib import Path

ROOT = Path(__file__).resolve().parent
VERDICTS = ROOT / ".." / "learning" / "media-corpus" / "verdicts.jsonl"

DOMINIOS_IMAGEN = {"image.pollinations.ai", "images.meigen.ai"}
PROVIDERS_AUDIO = {"edge-tts", "local", "composition"}
MOTIONS = {"zoom-in", "zoom-out", "pan-left", "pan-right", "pan-up", "pan-down"}
IDIOMAS = {"es", "en", "fr", "pt", "de", "it", "ar", "hi", "ja", "zh", "ru", "nl", "tr", "ko"}


def _url_ok(url: str) -> bool:
    return url.startswith(("https://", "/media/"))


def score_image(data: dict) -> tuple[int, list[str]]:
    """Max 25: url valida (10) + dominio conocido (10) + modelo (5)."""
    puntos = 0
    notas: list[str] = []
    url = data.get("url", "")
    if _url_ok(url):
        puntos += 10
    else:
        notas.append("url invalida")
    if any(url.startswith(d) or f"/{d.split('.')[0]}" in url for d in DOMINIOS_IMAGEN):
        puntos += 10
    elif url:
        notas.append("dominio no verificado")
    if data.get("model"):
        puntos += 5
    else:
        notas.append("sin modelo")
    return puntos, notas


def score_audio(data: dict) -> tuple[int, list[str]]:
    """Max 25: provider (10) + formato (10) + url (5)."""
    puntos = 0
    notas: list[str] = []
    if data.get("provider") in PROVIDERS_AUDIO:
        puntos += 10
    else:
        notas.append(f"provider '{data.get('provider')}' no esperado")
    url = data.get("url", "")
    if url and (url.endswith((".mp3", ".wav")) or _url_ok(url)):
        puntos += 10
    else:
        notas.append("sin archivo de audio")
    if _url_ok(url):
        puntos += 5
    else:
        notas.append("sin url servible")
    return puntos, notas


def score_video(data: dict) -> tuple[int, list[str]]:
    """Max 25: frames validos (10) o url (10) + motion (5) + provider (5)."""
    puntos = 0
    notas: list[str] = []
    frames = data.get("frames")
    url = data.get("url", "")
    if isinstance(frames, list) and 1 <= len(frames) <= 8 or _url_ok(url):
        puntos += 10
    else:
        notas.append("sin frames ni url")
    if data.get("motion") in MOTIONS:
        puntos += 5
    if data.get("provider"):
        puntos += 5
    return puntos, notas


def score_tts(data: dict) -> tuple[int, list[str]]:
    """Max 25: idioma (10) + voz edge-tts (10) + url mp3 (5)."""
    puntos = 0
    notas: list[str] = []
    if data.get("language") in IDIOMAS or data.get("language") == "multi":
        puntos += 10
    else:
        notas.append(f"idioma '{data.get('language')}' no soportado")
    voz = data.get("voz") or data.get("voice") or ""
    if "Neural" in str(voz):
        puntos += 10
    else:
        notas.append("voz edge-tts no reconocida")
    url = data.get("url", "")
    if url.endswith(".mp3") or _url_ok(url):
        puntos += 5
    else:
        notas.append("sin mp3")
    return puntos, notas


def score_director(data: dict) -> tuple[int, list[str]]:
    """Max 25: language (10) + script (10) + images>=1 (5)."""
    puntos = 0
    notas: list[str] = []
    plan = data.get("plan", data)
    if plan.get("language") in IDIOMAS or plan.get("language") == "multi":
        puntos += 10
    else:
        notas.append("language invalido")
    if plan.get("script"):
        puntos += 10
    else:
        notas.append("sin script")
    if isinstance(plan.get("images"), list) and len(plan["images"]) >= 1:
        puntos += 5
    else:
        notas.append("sin images")
    return puntos, notas


SCORERS = {
    "image": score_image,
    "audio": score_audio,
    "video": score_video,
    "tts": score_tts,
    "music": score_audio,
    "director": score_director,
}


def main() -> int:
    if len(sys.argv) < 2:
        print(__doc__)
        return 2
    verbose = "--verbose" in sys.argv
    data = json.loads(Path(sys.argv[1]).read_text(encoding="utf-8"))
    modalidad = data.get("modalidad", "image")
    scorer = SCORERS.get(modalidad, score_image)
    puntos, notas = scorer(data)
    ok = puntos >= 20  # >= 80% del maximo (25)
    veredicto = {
        "ts": "",
        "case_id": data.get("id", "media-unknown"),
        "modalidad": modalidad,
        "score": puntos,
        "max": 25,
        "status": "PASS" if ok else "FAIL",
        "notas": notas,
    }
    veredicto["ts"] = __import__("datetime").datetime.now(__import__("datetime").timezone.utc).isoformat()
    if VERDICTS.exists():
        with VERDICTS.open("a", encoding="utf-8") as f:
            f.write(json.dumps(veredicto, ensure_ascii=False) + "\n")
    print(f"[{'PASS' if ok else 'FAIL'}] {modalidad} {puntos}/25")
    if verbose and notas:
        for n in notas:
            print(f"  - {n}")
    return 0 if ok else 1


if __name__ == "__main__":
    sys.exit(main())

File: packages/core/src/tools/media-score.ts
/**
 * AutoPub F5 — media_score port a TypeScript (fuente:
 * ULTRAIA/integracionesImplementacion/media_score.py, criterios verificados).
 *
 * `puntuarMedia` puntúa un resultado de generación (imagen/audio/video/tts/music/director)
 * 0-25 con veredicto PASS (≥20) / FAIL. `puntuarPaquete` puntúa un PublicationPackage (F3)
 * 0-100 como score de calidad PRE-publicación (el plan maestro F5, tarea 2).
 */

const DOMINIOS_IMAGEN = ['image.pollinations.ai', 'images.meigen.ai'];
const PROVIDERS_AUDIO = ['edge-tts', 'local', 'composition'];
const MOTIONS = new Set(['zoom-in', 'zoom-out', 'pan-left', 'pan-right', 'pan-up', 'pan-down']);
const IDIOMAS = new Set(['es', 'en', 'fr', 'pt', 'de', 'it', 'ar', 'hi', 'ja', 'zh', 'ru', 'nl', 'tr', 'ko', 'multi']);

export type MediaModalidad = 'image' | 'audio' | 'video' | 'tts' | 'music' | 'director';

export interface MediaVeredicto {
  caseId: string;
  modalidad: MediaModalidad;
  score: number;
  max: number;
  status: 'PASS' | 'FAIL';
  notas: string[];
}

function urlOk(url: string): boolean {
  return url.startsWith('https://') || url.startsWith('/media/');
}

function scoreImage(data: Record<string, unknown>): [number, string[]] {
  let puntos = 0;
  const notas: string[] = [];
  const url = String(data.url ?? '');
  if (urlOk(url)) puntos += 10;
  else notas.push('url invalida');
  if (DOMINIOS_IMAGEN.some((d) => url.startsWith(d) || url.includes(`/${d.split('.')[0]}`))) puntos += 10;
  else if (url) notas.push('dominio no verificado');
  if (data.model) puntos += 5;
  else notas.push('sin modelo');
  return [puntos, notas];
}

function scoreAudio(data: Record<string, unknown>): [number, string[]] {
  let puntos = 0;
  const notas: string[] = [];
  const provider = String(data.provider ?? '');
  if (PROVIDERS_AUDIO.includes(provider)) puntos += 10;
  else notas.push(`provider '${provider}' no esperado`);
  const url = String(data.url ?? '');
  if (url && (url.endsWith('.mp3') || url.endsWith('.wav') || urlOk(url))) puntos += 10;
  else notas.push('sin archivo de audio');
  if (urlOk(url)) puntos += 5;
  else notas.push('sin url servible');
  return [puntos, notas];
}

function scoreVideo(data: Record<string, unknown>): [number, string[]] {
  let puntos = 0;
  const notas: string[] = [];
  const frames = data.frames;
  const url = String(data.url ?? '');
  if ((Array.isArray(frames) && frames.length >= 1 && frames.length <= 8) || urlOk(url)) puntos += 10;
  else notas.push('sin frames ni url');
  if (data.motion && MOTIONS.has(String(data.motion))) puntos += 5;
  if (data.provider) puntos += 5;
  return [puntos, notas];
}

function scoreTts(data: Record<string, unknown>): [number, string[]] {
  let puntos = 0;
  const notas: string[] = [];
  const language = String(data.language ?? '');
  if (IDIOMAS.has(language)) puntos += 10;
  else notas.push(`idioma '${language}' no soportado`);
  const voz = String(data.voz ?? data.voice ?? '');
  if (voz.includes('Neural')) puntos += 10;
  else notas.push('voz edge-tts no reconocida');
  const url = String(data.url ?? '');
  if (url.endsWith('.mp3') || urlOk(url)) puntos += 5;
  else notas.push('sin mp3');
  return [puntos, notas];
}

function scoreDirector(data: Record<string, unknown>): [number, string[]] {
  let puntos = 0;
  const notas: string[] = [];
  const plan = (data.plan ?? data) as Record<string, unknown>;
  const language = String(plan.language ?? '');
  if (IDIOMAS.has(language)) puntos += 10;
  else notas.push('language invalido');
  if (plan.script) puntos += 10;
  else notas.push('sin script');
  if (Array.isArray(plan.images) && plan.images.length >= 1) puntos += 5;
  else notas.push('sin images');
  return [puntos, notas];
}

const SCORERS: Record<MediaModalidad, (d: Record<string, unknown>) => [number, string[]]> = {
  image: scoreImage,
  audio: scoreAudio,
  video: scoreVideo,
  tts: scoreTts,
  music: scoreAudio,
  director: scoreDirector,
};

/** Puntúa un resultado de generación (port 1:1 de media_score.py). */
export function puntuarMedia(data: Record<string, unknown>): MediaVeredicto {
  const modalidad = (String(data.modalidad ?? 'image') || 'image') as MediaModalidad;
  const scorer = SCORERS[modalidad] ?? scoreImage;
  const [score, notas] = scorer(data);
  return {
    caseId: String(data.id ?? 'media-unknown'),
    modalidad,
    score,
    max: 25,
    status: score >= 20 ? 'PASS' : 'FAIL',
    notas,
  };
}

/** Paquete de publicación F3 (PublicationPackage) mínimo necesario para puntuar. */
export interface PaquetePuntual {
  caption?: string;
  hashtags?: string[] | string;
  visualByChannel?: Record<string, { formato?: string; thumbnail?: string } | undefined>;
  captionsByChannel?: Record<string, { caption?: unknown; srt?: unknown } | undefined>;
  contenido?: string;
  horarioSugerido?: string | Record<string, string>;
  canales?: string[];
  media?: string[];
  srtByChannel?: Record<string, string | undefined>;
}

/** Score 0-100 de calidad pre-publicación de un PublicationPackage. */
export function puntuarPaquete(paquete: PaquetePuntual): { score: number; notas: string[] } {
  let score = 0;
  const notas: string[] = [];
  const canales = paquete.canales ?? Object.keys(paquete.captionsByChannel ?? {});

  // 0-30: contenido
  if (paquete.contenido && paquete.contenido.trim().length >= 80) score += 30;
  else if (paquete.contenido && paquete.contenido.trim().length > 0) score += 15;
  else notas.push('contenido corto o vacio');

  // 0-20: caption + hashtags por canal
  const conCaption =
    canales.length > 0 &&
    canales.every((c) => {
      const caption = paquete.captionsByChannel?.[c]?.caption ?? paquete.caption;
      return typeof caption === 'string' && caption.trim().length >= 10;
    });
  if (conCaption) score += 20;
  else notas.push('falta caption por canal');

  const hashtags = Array.isArray(paquete.hashtags) ? paquete.hashtags : [];
  const conHashtags = hashtags.length >= 2;
  if (conHashtags) score += 10;
  else notas.push('menos de 2 hashtags');

  // 0-15: visual por canal (formato + thumbnail)
  const conVisual =
    canales.length > 0 &&
    canales.every((c) => {
      const v = paquete.visualByChannel?.[c];
      return v?.formato && v?.thumbnail;
    });
  if (conVisual) score += 15;
  else notas.push('falta visual/thumbnail por canal');

  // 0-10: subtitulos (video) o media (texto)
  const conSrt = canales.some((c) => paquete.srtByChannel?.[c] || paquete.captionsByChannel?.[c]?.srt);
  const conMedia = Array.isArray(paquete.media) && paquete.media.length > 0;
  if (conSrt || conMedia) score += 10;
  else notas.push('sin subtitulos ni media');

  // 0-15: horario sugerido (string o record por canal con al menos una entrada)
  const horario =
    typeof paquete.horarioSugerido === 'string'
      ? paquete.horarioSugerido
      : paquete.horarioSugerido
        ? Object.values(paquete.horarioSugerido).filter(Boolean)[0] ?? ''
        : '';
  if (horario) score += 15;
  else notas.push('sin horario sugerido');

  return { score: Math.min(score, 100), notas };
}

export const mediaScore = { puntuarMedia, puntuarPaquete };